Bug in Typescript definitions for Schema.index()

<!-- *Before creating an issue please make sure you are using the latest version of mongoose -->

**Do you want to request a *feature* or report a *bug*?**
bug (in Typescript types)

**What is the current behavior?**
The definition of Schema.index() (at https://github.com/Automattic/mongoose/blob/master/index.d.ts#L1279-L1280) says that the `fields` argument is of type `mongodb.IndexSpecification`, but mongoose actually takes its own shorthand syntax for the indexed fields, namely an object containing the indexed fields and the sort order (see https://mongoosejs.com/docs/api/schema.html#schema_Schema-index and https://github.com/Automattic/mongoose/blob/master/lib/schema.js#L1549-L1560).

**If the current behavior is a bug, please provide the steps to reproduce.**
<!-- If you can, provide a standalone script / gist to reproduce your issue. -->
This is as easy as the following schema definition (as a `.ts` file):

```ts
import mongoose from "mongoose";

export const schema = new mongoose.Schema({ whatever: String });

schema.index({ whatever: 1 });
```
... you'll get an error on the `whatever: 1` in the `schema.index()` call, stating:
```text
error TS2345: Argument of type '{ whatever: number; }' is not assignable to parameter of type 'IndexSpecification'.
```

<!-- If you are using TypeScript, please include your `tsconfig.json` -->
Here's the `tsconfig.json` file, though it shouldn't matter much:
```json
{
  "extends": "@tsconfig/node14/tsconfig.json",

  "compilerOptions": {
    "rootDir": "./",
    "outDir": "./dist",

    "allowJs": true,
    "checkJs": false,

    "typeRoots": [
      "@types",
      "node_modules/@types"
    ]
  },

  "include": ["**/*"]
}
```


**What is the expected behavior?**
No error, the `fields` argument to `schema.index()` should probably have a type definition similar to `SchemaDefinition<>`/`DocumentDefinition<>` that leverages the `SchemaDefinitionType` to define the allowable keys.  Maybe something like
```ts
export type IndexDefinition<T> = { [K in keyof T]: 1 | -1 };
// ...

/** Defines an index (most likely compound) for this schema. */
index(fields: IndexDefinition<DocumentDefinition<SchemaDefinitionType>>>, options?: IndexOptions): this;
```

(The `SchemaDefinition<>` type might need to be in that chain as well... it's not clear to me what specific value that layer adds.)

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node.js: 14.16.0
mongoose: 5.13.6
mongodb: irrelevant

<!-- You can print `mongoose.version` to get your current version of Mongoose: https://mongoosejs.com/docs/api.html#mongoose_Mongoose-version -->

--- a/index.d.ts
+++ b/index.d.ts
@@ -1246,7 +1246,8 @@ declare module 'mongoose' {
   type ExtractQueryHelpers<M> = M extends Model<any, infer TQueryHelpers> ? TQueryHelpers : {};
   type ExtractMethods<M> = M extends Model<any, any, infer TMethods> ? TMethods : {};
 
-  type IndexDefinition<T> = { [K in keyof T]: 1 | -1; };
+  type IndexDirection = 1 | -1;
+  type IndexDefinition<T> = { [K in keyof T]: IndexDirection; };
 
   type PreMiddlewareFunction<T> = (this: T, next: (err?: CallbackError) => void) => void | Promise<void>;
   type PreSaveMiddlewareFunction<T> = (this: T, next: (err?: CallbackError) => void, opts: SaveOptions) => void | Promise<void>;

--- a/index.d.ts
+++ b/index.d.ts
@@ -1246,6 +1246,8 @@ declare module 'mongoose' {
   type ExtractQueryHelpers<M> = M extends Model<any, infer TQueryHelpers> ? TQueryHelpers : {};
   type ExtractMethods<M> = M extends Model<any, any, infer TMethods> ? TMethods : {};
 
+  type IndexDefinition<T> = { [K in keyof T]: 1 | -1; };
+
   type PreMiddlewareFunction<T> = (this: T, next: (err?: CallbackError) => void) => void | Promise<void>;
   type PreSaveMiddlewareFunction<T> = (this: T, next: (err?: CallbackError) => void, opts: SaveOptions) => void | Promise<void>;
   type PostMiddlewareFunction<ThisType, ResType = any> = (this: ThisType, res: ResType, next: (err?: CallbackError) => void) => void | Promise<void>;
@@ -1277,13 +1279,13 @@ declare module 'mongoose' {
     eachPath(fn: (path: string, type: SchemaType) => void): this;
 
     /** Defines an index (most likely compound) for this schema. */
-    index(fields: mongodb.IndexSpecification, options?: IndexOptions): this;
+    index(fields: IndexDefinition<SchemaDefinition<DocumentDefinition<SchemaDefinitionType>>>, options?: IndexOptions): this;
 
     /**
      * Returns a list of indexes that this schema declares, via `schema.index()`
      * or by `index: true` in a path's options.
      */
-    indexes(): Array<any>;
+    indexes(): Array<IndexDefinition<SchemaDefinition<DocumentDefinition<SchemaDefinitionType>>>>;
 
     /** Gets a schema option. */
     get<K extends keyof SchemaOptions>(key: K): SchemaOptions[K];

--- a/index.d.ts
+++ b/index.d.ts
@@ -1246,9 +1246,9 @@ declare module 'mongoose' {
   type ExtractQueryHelpers<M> = M extends Model<any, infer TQueryHelpers> ? TQueryHelpers : {};
   type ExtractMethods<M> = M extends Model<any, any, infer TMethods> ? TMethods : {};
 
-  type IndexDirection = 1 | -1;
-  type IndexDefinition<T> = { [K in keyof T]: IndexDirection; };
-
+  type IndexDirection = 1 | -1 | '2d' | '2dsphere' | 'geoHaystack' | 'hashed' | 'text';
+  type IndexDefinition = Record<string, IndexDirection>;
+  
   type PreMiddlewareFunction<T> = (this: T, next: (err?: CallbackError) => void) => void | Promise<void>;
   type PreSaveMiddlewareFunction<T> = (this: T, next: (err?: CallbackError) => void, opts: SaveOptions) => void | Promise<void>;
   type PostMiddlewareFunction<ThisType, ResType = any> = (this: ThisType, res: ResType, next: (err?: CallbackError) => void) => void | Promise<void>;
@@ -1280,13 +1280,13 @@ declare module 'mongoose' {
     eachPath(fn: (path: string, type: SchemaType) => void): this;
 
     /** Defines an index (most likely compound) for this schema. */
-    index(fields: IndexDefinition<SchemaDefinition<DocumentDefinition<SchemaDefinitionType>>>, options?: IndexOptions): this;
+    index(fields: IndexDefinition, options?: IndexOptions): this;
 
     /**
      * Returns a list of indexes that this schema declares, via `schema.index()`
      * or by `index: true` in a path's options.
      */
-    indexes(): Array<IndexDefinition<SchemaDefinition<DocumentDefinition<SchemaDefinitionType>>>>;
+    indexes(): Array<IndexDefinition>;
 
     /** Gets a schema option. */
     get<K extends keyof SchemaOptions>(key: K): SchemaOptions[K];

